### PR TITLE
Resolve issue 362

### PR DESCRIPTION
**Pull request by andrewgordstewart**
Right now, only one action is routed and only when it's player A.

The reason for restricting to player A is that the test scenarios aren't built out yet for player B, which I would prefer to do as a separate piece of work. At that point, both the code and the test can be copy-pasted from player A.

### CODE REVIEW

File: packages/wallet/src/redux/reducer.ts [hunk 1]
```diff
@@ -63,9 +65,25 @@ function routeToProtocolReducer(state, action) {
...
-function routeToNewProcessInitializer(state, action) {
-  // TODO: Call the process creator
-  return state;
+function routeToNewProcessInitializer(
+  state: states.Initialized,
+  action: actions.protocol.NewProcessAction,
+) {
+  switch (action.type) {
+    case actions.indirectFunding.FUNDING_REQUESTED:
+      switch (action.playerIndex) {
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/reducer.ts]: Actually, this logic should live in the `indirect-funding` protocol.

File: packages/wallet/src/redux/__tests__/initialized.test.ts [hunk 2]
```diff
@@ -26,3 +31,12 @@ describe('when the player initializes a channel', () => {
...
   });
 });
+
+describe('when a NewProcessAction arrives', () => {
+  const action = actions.indirectFunding.fundingRequested(channelId, PlayerIndex.A);
+  const initialize = jest.fn();
+  Object.defineProperty(IndirectFunding, 'initialize', { value: initialize });
+
+  walletReducer(initializedState, action);
+  expect(initialize).toHaveBeenCalledWith(action, states.EMPTY_SHARED_DATA);
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/__tests__/initialized.test.ts]: I need to expect something about the updated wallet state's `processStore`

File: packages/wallet/src/redux/reducer.ts [hunk 3]
```diff
@@ -63,9 +65,23 @@ function routeToProtocolReducer(state, action) {
...
+    case actions.indirectFunding.FUNDING_REQUESTED:
+      const { protocolState, sharedData } = indirectFunding.initialize(
+        action,
+        states.sharedData(state),
+      );
+
+      return startProcess(state, sharedData, action, protocolState);
+    default:
+      return state;
+    // TODO: Why is the discriminated union not working here?
```
Comment (lalexgap) [inline packages/wallet/src/redux/reducer.ts]: Weird, what was the error from `unreachable` ?

File: packages/wallet/src/redux/reducer.ts [hunk 4]
```diff
@@ -90,3 +106,25 @@ const waitForLoginReducer = (
       return state;
   }
 };
+function startProcess(
+  state: states.Initialized,
+  sharedData: states.SharedData,
+  action: {
```
Comment (lalexgap) [inline packages/wallet/src/redux/reducer.ts]: Is there a reason we're not just using a `FundingRequested` action type?

File: packages/wallet/src/redux/reducer.ts [hunk 5]
```diff
@@ -90,3 +106,25 @@ const waitForLoginReducer = (
...
+    type: 'WALLET.INDIRECT_FUNDING.FUNDING_REQUESTED';
+    channelId: string;
+    playerIndex: PlayerIndex;
+    protocol: WalletProtocol;
+  },
+  protocolState:
+    | states.indirectFunding.playerA.WaitForApproval
```
Comment (lalexgap) [inline packages/wallet/src/redux/reducer.ts]: Could we type this more broadly? Like `ProtocolState`?

It seems strange to have a `startProcess` method expecting  just `IndirectFunding` arguments. It seems like it should take `ProtocolState` and a `ProtocolAction` even if thats only happens to be indirect funding stuff right now.

File: packages/wallet/src/redux/reducer.ts (hunk 3 above)
Comment (andrewgordstewart) [reply packages/wallet/src/redux/reducer.ts]:
```
Argument of type '{ type: "WALLET.INDIRECT_FUNDING.FUNDING_REQUESTED"; channelId: string; playerIndex: PlayerIndex; protocol: WalletProtocol; }' is not assignable to parameter of type 'never'.
```

?

File: packages/wallet/src/redux/reducer.ts (hunk 5 above)
Comment (andrewgordstewart) [reply packages/wallet/src/redux/reducer.ts]: Yeah, thanks for catching that. It's from a VSCode refactor that I neglected to update.

File: packages/wallet/src/redux/reducer.ts (hunk 3 above)
Comment (tomclose) [reply packages/wallet/src/redux/reducer.ts]: I've had this problem when there's only one type before. I think if there's only one possibility for a type, typescript doesn't bother narrowing it in the switch statement.

File: packages/wallet/src/redux/reducer.ts [hunk 6]
```diff
@@ -90,3 +135,26 @@ const waitForLoginReducer = (
       return state;
   }
 };
+function startProcess(
+  state: states.Initialized,
+  sharedData: states.SharedData,
```
Comment (tomclose) [inline packages/wallet/src/redux/reducer.ts]: Doesn't the `state` contain `sharedData`?

File: packages/wallet/src/redux/state.ts [hunk 7]
```diff
@@ -71,27 +73,27 @@ export function indirectFundingOngoing(state: Initialized): state is IndirectFun
...
 // ------------
-export const emptyState: Shared = {
+export const EMPTY_SHARED_DATA: SharedData = {
   outboxState: emptyDisplayOutboxState(),
-  channelState: { initializedChannels: {}, initializingChannels: {} },
+  channelState: emptyChannelState(),
 };
 
-function shared(params: Shared): Shared {
+export function sharedData(params: SharedData): SharedData {
```
Comment (tomclose) [inline packages/wallet/src/redux/state.ts]: Unrelated to this PR, but isn't this function pointless?

**Pull request merged**